presto-parquet/src/main/java/com/facebook/presto/parquet/cache/MetadataReader.java

```diff
@@ -112,9 +126,52 @@ public static ParquetFileMetadata readFooter(ParquetDataSource parquetDataSource
+        return readParquetMetadata(tailSlice.slice(tailSlice.length() - completeFooterSize, metadataLength).getInput(), metadataLength, modificationTime, fileDecryptor, encryptedFooterMode, parquetDataSource.getId());
+    }
+
+    private static ParquetFileMetadata readParquetMetadata(BasicSliceInput input, int metadataLength, long modificationTime, Optional<InternalFileDecryptor> fileDecryptor, boolean encryptedFooterMode, ParquetDataSourceId id)
+            throws IOException
+    {
+        if (encryptedFooterMode && !fileDecryptor.isPresent()) {
+            new IllegalArgumentException("fileDecryptionProperties cannot be null when encryptedFooterMode is true");
+        }
+        Decryptor footerDecryptor = null;
+        byte[] aad = null;
+
+        if (encryptedFooterMode) {
+            FileCryptoMetaData fileCryptoMetaData = readFileCryptoMetaData(input);
+            fileDecryptor.get().setFileCryptoMetaData(fileCryptoMetaData.getEncryption_algorithm(), true, fileCryptoMetaData.getKey_metadata());
+            footerDecryptor = fileDecryptor.get().fetchFooterDecryptor();
+            aad = AesCipher.createFooterAAD(fileDecryptor.get().getFileAAD());
+        }
+
+        FileMetaData fileMetaData = readFileMetaData(input, footerDecryptor, aad);
+        return convertToParquetMetadata(input, fileMetaData, metadataLength, modificationTime, fileDecryptor, encryptedFooterMode, id);
+    }
+
+    private static ParquetFileMetadata convertToParquetMetadata(BasicSliceInput input, FileMetaData fileMetaData, int metadataLength, long modificationTime, Optional<InternalFileDecryptor> fileDecryptor, boolean encryptedFooter, ParquetDataSourceId id)
+            throws IOException
+    {
         List<SchemaElement> schema = fileMetaData.getSchema();
-        validateParquet(!schema.isEmpty(), "Empty Parquet schema in file: %s", parquetDataSource.getId());
+        if (!encryptedFooter && fileDecryptor.isPresent()) {
+            if (!fileMetaData.isSetEncryption_algorithm()) { // Plaintext file
+                fileDecryptor.get().setPlaintextFile();
+                // Done to detect files that were not encrypted by mistake
+                if (!fileDecryptor.get().plaintextFilesAllowed()) {
+                    throw new ParquetCryptoRuntimeException("Applying decryptor on plaintext file");
+                }
+            }
+            else {  // Encrypted file with plaintext footer
+                // if no fileDecryptor, can still read plaintext columns
+                fileDecryptor.get().setFileCryptoMetaData(fileMetaData.getEncryption_algorithm(), false,
+                        fileMetaData.getFooter_signing_key_metadata());
+                if (fileDecryptor.get().checkFooterIntegrity()) {
+                    verifyFooterIntegrity(input, fileDecryptor.get(), metadataLength);
+                }
+            }
+        }
+
+        validateParquet(!schema.isEmpty(), "Empty Parquet schema in file: %s", id);
```

Comment:
keep this line before encryptedFooter and fileDecryptor check

Reply:
Good catch!